optaplanner-examples/src/main/java/org/optaplanner/examples/scrabble/optional/score/ScrabbleConstraintProvider.java

```diff
@@ -0,0 +1,69 @@
+package org.optaplanner.examples.scrabble.optional.score;
+
+import org.optaplanner.core.api.score.buildin.hardmediumsoft.HardMediumSoftScore;
+import org.optaplanner.core.api.score.stream.Constraint;
+import org.optaplanner.core.api.score.stream.ConstraintCollectors;
+import org.optaplanner.core.api.score.stream.ConstraintFactory;
+import org.optaplanner.core.api.score.stream.ConstraintProvider;
+import org.optaplanner.core.api.score.stream.Joiners;
+import org.optaplanner.examples.scrabble.domain.ScrabbleWordDirection;
+import org.optaplanner.examples.scrabble.domain.ScrabbleCell;
+import org.optaplanner.examples.scrabble.domain.ScrabbleWordAssignment;
+
+public class ScrabbleConstraintProvider implements ConstraintProvider {
+
+    @Override
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[]{
+                                characterConflict(constraintFactory),
+                                noParallelHorizontalNeighbours(constraintFactory),
+                                noParallelVerticalNeighbours(constraintFactory),
+                                outOfGrid(constraintFactory),
+                                maximizeMergesPerWord(constraintFactory),
+                                pullToCenter(constraintFactory)
+        };
+    }
+
+    private Constraint characterConflict(ConstraintFactory cf) {
+        return cf.from(ScrabbleCell.class)
+                 .filter(sc -> sc.getCharacterSet().size() >= 2)
+                 .penalize("Character confict", HardMediumSoftScore.ONE_HARD, sc -> sc.getCharacterSet().size() - 1);
+    }
+
+    private Constraint noParallelHorizontalNeighbours(ConstraintFactory cf) {
+        return cf.from(ScrabbleCell.class).filter(sc -> sc.hasWordSet(ScrabbleWordDirection.HORIZONTAL))
+                 .join(ScrabbleCell.class,
+                       Joiners.equal(ScrabbleCell::getX), Joiners.lessThan(ScrabbleCell::getId),
+                       Joiners.filtering((first, other) -> other.hasWordSet(ScrabbleWordDirection.HORIZONTAL)))
+                 .filter((c1, c2) -> Math.abs(c1.getY() - c2.getY()) == 1)
+                 .penalize("No parallel horizontal neighbours", HardMediumSoftScore.ONE_HARD);
+    }
+
+    private Constraint noParallelVerticalNeighbours(ConstraintFactory cf) {
+        return cf.from(ScrabbleCell.class).filter(sc -> sc.hasWordSet(ScrabbleWordDirection.VERTICAL))
+                 .join(ScrabbleCell.class,
+                       Joiners.equal(ScrabbleCell::getY), Joiners.lessThan(ScrabbleCell::getId),
+                       Joiners.filtering((first, other) -> other.hasWordSet(ScrabbleWordDirection.VERTICAL)))
+                 .filter((c1, c2) -> Math.abs(c1.getX() - c2.getX()) == 1)
+                 .penalize("No parallel vertical neighbours", HardMediumSoftScore.ONE_HARD);
+    }
```

Comment:
Looking at the code now, the `join().filter()` could be replace by `ifExists(...)` with the appropriate joiners. The performance of that would likely be the best - only not sure if it would match 1:1 with what the DRL does. If FULL_ASSERT passes, I say let's do it.

Reply:
I don't think they are the same. Consider this scenario:

```
WORD1
WORD2
WORD3
```

Where `id(WORD2) < id(WORD1) < id(WORD3)`.
Using the current method, the constraint matches are `[(WORD2, WORD1),(WORD2,WORD3)]`.
Using the `ifExists` method, the constraint matches are `[(WORD2,WORD1|WORD2)]` (there is only one constraint match, with the second element being either WORD1 or WORD2). This situation is extremely rare, and probably take a long time to be explored. I ran 20 minutes on FULL_ASSERT using `ifExists` and got no errors so far, but I still think it not one-to-one.

Reply:
If (`cell1.x < cell2.x -> cell1.id < cell2.id` and `cell1.y < cell2.y -> cell1.id < cell2.id`) or (`cell1.x > cell2.x -> cell1.id > cell2.id` and `cell1.y > cell2.y -> cell1.id > cell2.id`), this could never happen (it requires an id discontinuity)

Reply:
From the importer:
```
        private void readCellList() throws IOException {
            int gridWidth = readIntegerValue("gridWidth:");
            solution.setGridWidth(gridWidth);
            int gridHeight = readIntegerValue("gridHeight:");
            solution.setGridHeight(gridHeight);
            List<ScrabbleCell> cellList = new ArrayList<>(gridWidth * gridHeight);
            for (int x = 0; x < gridWidth; x++) {
                for (int y = 0; y < gridHeight; y++) {
                    ScrabbleCell cell = new ScrabbleCell();
                    cell.setId((long) (y * gridWidth + x));
                    cell.setX(x);
                    cell.setY(y);
                    cell.setWordSet(new LinkedHashSet<>());
                    cell.setCharacterCountMap(new LinkedHashMap<>());
                    cellList.add(cell);
                }
            }
            solution.setCellList(cellList);
        }
```
In particular: `cell.setId((long) (y * gridWidth + x))`
Due to this, the corner case will never be hit, and the constraints are "equivalent".
"Proof":
Assume, for contradiction, there exists cells c1, c2, c3 such that `c1.x = c2.x = c3.x` and `c1.hasWordSet(HORIZONTAL) and c2.hasWordSet(HORIZONTAL) and c3.hasWordSet(HORIZONTAL)`, with `c1.y < c2.y < c3.y`. From the id formula we can derive:
```
c1.id 
= c1.y * W + x
< c2.y * W + x
= c2.id
```
Thus `c1.id < c2.id`. Similarly we have:
```
c2.id
= c2.y * W + x
< c3.y * W + x
= c3.id
```
Thus `c2.id < c3.id` and we have `c1.id < c2.id < c3.id`. This make "double" matches impossible, as they can only occur if a cell is between two cells with an id less than both of them, which we proved is not possible. QED.

Reply:
Side note: This proof also proves the filter can be simplified to `c1.y == c2.y - 1`.

Reply:
@triceo if Id generating code won't change and no one using Scrabble in production (with their own id generating code that might not have this property that cause the constraints to be equivalent), then I am okay changing it to `ifExists`.

Reply:
Looking at the `Drools`, they indeed use `c1.x = c2.x and c2.y = c1.y + 1`, meaning we can safety change this to an `ifExists` and remove the id joiner (which turn out was only changing `Math.abs(c1.y - c2.y) = 1` to `c2.y = c1.y + 1`).

Reply:
(To bad there isn't a `Joiners.fixed((first,second) -> boolean)`, since the cells `x` and `y` don't ever change, meaning we already know what cell we need to check for a particular cell and don't need to check all the cells in the same column)

Reply:
(`fixed` is basically a filtering that is only run once)

Reply:
(wait, `equals` have an override that use two different functions; I can use that)

Reply:
Well proven!